File: solutions.digamma.damas.jcr/src/main/java/solutions/digamma/damas/jcr/model/JcrReadManager.java
package solutions.digamma.damas.jcr.model;

import solutions.digamma.damas.common.WorkspaceException;
import solutions.digamma.damas.entity.Entity;
import solutions.digamma.damas.entity.EntityManager;
import solutions.digamma.damas.auth.Token;
import solutions.digamma.damas.inspection.NotNull;
import solutions.digamma.damas.jcr.common.Exceptions;
import solutions.digamma.damas.jcr.session.SessionWrapper;
import solutions.digamma.damas.logging.Logged;

import javax.jcr.RepositoryException;
import javax.jcr.Session;

/**
 * Abstract read manager.
 *
 * @author Ahmad Shahwan
 */
abstract public class JcrReadManager<T extends Entity>
        extends JcrManager implements EntityManager<T> {

    @Logged
    @Override
    public T retrieve(@NotNull Token token, @NotNull String id)
            throws WorkspaceException {
        try (SessionWrapper session = openSession(token)) {
            return this.retrieve(session.getSession(), id);
        } catch (RepositoryException e) {
            throw Exceptions.convert(e);
        }
    }

    /**
     * Perform retrieval.
     *
     * @param session
     * @param id
     * @return
     * @throws RepositoryException
     * @throws WorkspaceException
     */
    abstract protected T retrieve(@NotNull Session session, @NotNull String id)
            throws RepositoryException, WorkspaceException;
}

File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/model/JcrReadManager.kt
package solutions.digamma.damas.jcr.model

import solutions.digamma.damas.common.WorkspaceException
import solutions.digamma.damas.entity.Entity
import solutions.digamma.damas.entity.EntityManager
import solutions.digamma.damas.jcr.common.Exceptions
import solutions.digamma.damas.logging.Logged
import javax.jcr.RepositoryException

/**
 * Abstract read manager.
 *
 * @author Ahmad Shahwan
 */
internal abstract class JcrReadManager<T: Entity> : JcrManager(),
        EntityManager<T> {

    /**
     * Retrieve an existing entity.
     *
     * This method converts a thrown exception into a [WorkspaceException].
     */
    @Logged
    @Throws(WorkspaceException::class)
    override fun retrieve(id: String): T = Exceptions.check {
        this.doRetrieve(id)
    }

    /**
     * Perform retrieval.
     *
     * @param id
     * @return
     * @throws RepositoryException
     * @throws WorkspaceException
     */
    @Throws(RepositoryException::class, WorkspaceException::class)
    protected abstract fun doRetrieve(id: String): T
}
